src/pal/src/numa/numa.cpp

```diff
@@ -0,0 +1,683 @@
+#endif
+
+int GetNumberOfProcessors();
+
+// CPU affinity descriptor
+struct CpuAffinity
+{
+    // NUMA node
+    BYTE Node;
+    // CPU number relative to the group the CPU is in
+    BYTE Number;
+    // CPU group
+    WORD Group;
+};
+
+// Array mapping global CPU index to its affinity
+CpuAffinity *g_cpuToAffinity = NULL;
+
+// Array mapping CPU group and index in the group to the global CPU index
+short *g_groupAndIndexToCpu = NULL;
+// Array mapping CPU group to the corresponding affinity mask of the CPUs in the group
+KAFFINITY *g_groupToCpuMask = NULL;
+// Array mapping CPU group to the number of processors in the group
+BYTE *g_groupToCpuCount = NULL;
+
+// Total number of processors in the system
+int g_cpuCount = 0;
+// Total number of CPU groups
+int g_groupCount = 0;
+// The highest NUMA node available
+int g_highestNumaNode = 0;
+
+static const int MaxCpusPerGroup = 8 * sizeof(KAFFINITY);
+static const WORD NO_GROUP = 0xffff;
+
+/*++
+Function:
+  FreeLookupArrays
+
+Allocate CPU and group lookup arrays
+--*/
+VOID
+AllocateLookupArrays()
+{
+    g_groupAndIndexToCpu = (short*)malloc(g_groupCount * MaxCpusPerGroup * sizeof(short));
+    g_cpuToAffinity = (CpuAffinity*)malloc(g_cpuCount * sizeof(CpuAffinity));
+    g_groupToCpuMask = (KAFFINITY*)malloc(g_groupCount * sizeof(KAFFINITY));
+    g_groupToCpuCount = (BYTE*)malloc(g_groupCount * sizeof(BYTE));
+
+    memset(g_groupAndIndexToCpu, 0xff, g_groupCount * MaxCpusPerGroup * sizeof(short));
+    memset(g_cpuToAffinity, 0xff, g_cpuCount * sizeof(CpuAffinity));
+    memset(g_groupToCpuMask, 0, g_groupCount * sizeof(KAFFINITY));
+    memset(g_groupToCpuCount, 0, g_groupCount * sizeof(BYTE));
+}
+
+/*++
+Function:
+  FreeLookupArrays
+
+Free CPU and group lookup arrays
+--*/
+VOID
+FreeLookupArrays()
+{
+    free(g_groupAndIndexToCpu);
+    free(g_cpuToAffinity);
+    free(g_groupToCpuMask);
+    free(g_groupToCpuCount);
+
+    g_groupAndIndexToCpu = NULL;
+    g_cpuToAffinity = NULL;
+    g_groupToCpuMask = NULL;
+    g_groupToCpuCount = NULL;
+}
+
+/*++
+Function:
+  GetFullAffinityMask
+
+Get affinity mask for the specified number of processors with all
+the processors enabled.
+--*/
+KAFFINITY GetFullAffinityMask(int cpuCount)
+{
+    return ((KAFFINITY)1 << (cpuCount)) - 1;
+}
+
+/*++
+Function:
+  NUMASupportInitialize
+
+Initialize data structures for getting and setting thread affinities to processors and
+querying NUMA related processor information.
+On systems with no NUMA support, it behaves as if there was a single NUMA node with
+a single group of processors.
+--*/
+BOOL
+NUMASupportInitialize()
+{
+#if HAVE_NUMA_H
+    if (numa_available() != -1)
+    {
+        struct bitmask *mask = numa_allocate_cpumask();
+        int numaNodesCount = numa_max_node() + 1;
+
+        g_cpuCount = numa_num_possible_cpus();
+        g_groupCount = 0;
+
+        for (int i = 0; i < numaNodesCount; i++)
+        {
+            int st = numa_node_to_cpus(i, mask);
```

Comment:
Adding assert, the only failure of this function could be that the mask is not large enough, but since the mask is allocated by the numa_allocate_cpumask function, that cannot happen.